drivers/adc/ad463x/ad463x.c

```diff
@@ -271,14 +271,14 @@ int32_t ad463x_exit_reg_cfg_mode(struct ad463x_dev *dev)
 				   AD463X_EXIT_CFG_MODE);
 	if (ret != 0)
 		return ret;
-
+#ifndef USE_STANDARD_SPI
```

Comment:
i think you could do something like has been done here to avoid these macros (see my first comment and the discussion that followed)
https://github.com/analogdevicesinc/no-OS/pull/2248

Reply:
sure, i prefer not to use macros too. 
It just means that the spi engine stuff will be built and not used
when using standard SPI for all platforms (including stm32).. 

ok with that?

Reply:
yes, i'm ok with that

Reply:
not so trivial to do. im starting to run into  issues when compiling for STM32:

* spi engine is using  #include "xilinx_spi.h"
so i have to add a hacky  (DRIVERS)/platform/xilinx/xilinx_spi.h"
to the src.mk. not big deal

but then:
* spi_engine is using usleep that is not available in STM32:
/drivers/axi_core/spi_engine/spi_engine.c :
#include <sleep.h>

need to start fixing these to see if i get to compile ;(

Reply:
seems to be the only error, so just switching to no_os_delay, in spi_engine looks like enough.

Reply:
but then, we get into missing all the axi_ funtions, so i guess we would have to provide
spi_engine stubs for other platforms?


```
/opt/st/stm32cubeide_1.14.0/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.11.3.rel1.linux64_1.1.100.202309141235/tools/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: /home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/objs/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/axi_dmac/axi_dmac.o: in function `axi_dmac_read':
/home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/axi_dmac/axi_dmac.c:236: undefined reference to `no_os_axi_io_read'
/opt/st/stm32cubeide_1.14.0/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.11.3.rel1.linux64_1.1.100.202309141235/tools/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: /home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/objs/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/axi_dmac/axi_dmac.o: in function `axi_dmac_write':
/home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/axi_dmac/axi_dmac.c:254: undefined reference to `no_os_axi_io_write'
/opt/st/stm32cubeide_1.14.0/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.11.3.rel1.linux64_1.1.100.202309141235/tools/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: /home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/objs/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/clk_axi_clkgen/clk_axi_clkgen.o: in function `axi_clkgen_write':
/home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/clk_axi_clkgen/clk_axi_clkgen.c:165: undefined reference to `no_os_axi_io_write'
/opt/st/stm32cubeide_1.14.0/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.11.3.rel1.linux64_1.1.100.202309141235/tools/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: /home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/objs/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/clk_axi_clkgen/clk_axi_clkgen.o: in function `axi_clkgen_read':
/home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/clk_axi_clkgen/clk_axi_clkgen.c:177: undefined reference to `no_os_axi_io_read'
/opt/st/stm32cubeide_1.14.0/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.11.3.rel1.linux64_1.1.100.202309141235/tools/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: /home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/objs/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/spi_engine/spi_engine.o: in function `spi_engine_write':
/home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/spi_engine/spi_engine.c:98: undefined reference to `no_os_axi_io_write'
/opt/st/stm32cubeide_1.14.0/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.11.3.rel1.linux64_1.1.100.202309141235/tools/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: /home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/objs/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/spi_engine/spi_engine.o: in function `spi_engine_read':
/home/axelh/gits/no-OS2/projects/ad463x_fmcz/build/app/Core/noos/drivers/axi_core/spi_engine/spi_engine.c:116: undefined reference to `no_os_axi_io_read'
```

Reply:
the stubs approach looks good, would marking them with weak linkage be of any help ?
like these for example: https://github.com/analogdevicesinc/no-OS/blob/main/util/no_os_mutex.c#L47

Reply:
@buha 

i could probably make it work with stubs too, but we might have the additional burden in the 
projects to conditionally compile the .c file. the way its right now, the project does not need to do anything special the implementation selection is taken into account by the STANDARD_SPI flag in spi_engine.c

i think weak is for defining multiple implementations of a function, and allow for future unknown implementations too. 
in this case  we only have two known implementations: do-nothing and do-something.

Reply:
you're right